test/e2e/annotations/fastcgi.go

```diff
@@ -0,0 +1,130 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package annotations
+
+import (
+	"net/http"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/parnurzeal/gorequest"
+	corev1 "k8s.io/api/core/v1"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/ingress-nginx/test/e2e/framework"
+)
+
+var _ = framework.IngressNginxDescribe("Annotations - FastCGI", func() {
+	f := framework.NewDefaultFramework("fastcgi")
+
+	BeforeEach(func() {
+		f.NewFastCGIHelloServerDeployment()
+	})
+
+	It("should use fastcgi_pass in the configuration file", func() {
+		host := "fastcgi"
+
+		annotations := map[string]string{
+			"nginx.ingress.kubernetes.io/backend-protocol": "FCGI",
+		}
+
+		ing := framework.NewSingleIngress(host, "/hello", host, f.Namespace, "fastcgi-helloserver", 9000, &annotations)
+		f.EnsureIngress(ing)
+
+		f.WaitForNginxServer(host,
+			func(server string) bool {
+				return Expect(server).Should(ContainSubstring("include /etc/nginx/fastcgi_params;")) &&
+					Expect(server).Should(ContainSubstring("fastcgi_pass"))
+			})
+	})
+
+	It("should add fastcgi_index in the configuration file", func() {
+		host := "fastcgi-index"
+
+		annotations := map[string]string{
+			"nginx.ingress.kubernetes.io/backend-protocol": "FCGI",
+			"nginx.ingress.kubernetes.io/fastcgi-index":    "index.php",
+		}
+
+		ing := framework.NewSingleIngress(host, "/hello", host, f.Namespace, "fastcgi-helloserver", 9000, &annotations)
+		f.EnsureIngress(ing)
+
+		f.WaitForNginxServer(host,
+			func(server string) bool {
+				return Expect(server).Should(ContainSubstring("fastcgi_index \"index.php\";"))
+			})
+	})
+
+	It("should add fastcgi_param in the configuration file", func() {
+		configuration := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "fastcgi-configmap",
+				Namespace: f.Namespace,
+			},
+			Data: map[string]string{
+				"SCRIPT_FILENAME": "/home/www/scripts/php$fastcgi_script_name",
+				"REDIRECT_STATUS": "200",
+			},
+		}
+
+		cm, err := f.EnsureConfigMap(configuration)
+		Expect(err).NotTo(HaveOccurred(), "failed to create an the configmap")
+		Expect(cm).NotTo(BeNil(), "expected a configmap but none returned")
+
+		host := "fastcgi-params-configmap"
+
+		annotations := map[string]string{
+			"nginx.ingress.kubernetes.io/backend-protocol":         "FCGI",
+			"nginx.ingress.kubernetes.io/fastcgi-params-configmap": "fastcgi-configmap",
+		}
+
+		ing := framework.NewSingleIngress(host, "/hello", host, f.Namespace, "fastcgi-helloserver", 9000, &annotations)
+		f.EnsureIngress(ing)
+
+		f.WaitForNginxServer(host,
+			func(server string) bool {
+				return Expect(server).Should(ContainSubstring("fastcgi_param SCRIPT_FILENAME \"/home/www/scripts/php$fastcgi_script_name\";")) &&
+					Expect(server).Should(ContainSubstring("fastcgi_param REDIRECT_STATUS \"200\";"))
```

Comment:
Would it make sense to include the value of these env vars in response and assert on response (I realize that requires modifying the fastcgi hello world Go app)? IMHO that would be more "end-to-end" than asserting on configuration directive.

Reply:
The last test in this file is the one which does the complete end-to-end proof. We did not wish to test all the annotations end-to-end because it made the test more fragile since it also tests nginx logic. Therefore, we did one full end-to-end test and the others have a slightly smaller scope, only testing that we are generating the correct configuration. If you prefer we can modify the hello image to include those variable to have every test be end-to-end.

Reply:
I don't feel too strongly about this.